### PR TITLE
seed: ReadSystemEssentialAndBetterEarliestTime

### PR DESCRIPTION
**Pull request by pedronis**
ReadSystemEssentialAndBetterEarliestTime retrieves in one go
information about the model and essential snaps of the given types
for the Core 20 recovery system seed specified by seedDir and label
(which cannot be empty).
It can operate even if current system time is unreliable by taking
a earliestTime lower bound for current time.
It returns as well an improved lower bound by considering appropriate
assertions in the seed.

### CODE REVIEW

**Comment by pedronis** [comment]
based on #10004

File: seed/seed.go [hunk 1]
```diff
@@ -153,3 +154,62 @@ func ReadSystemEssential(seedDir, label string, essentialTypes []snap.Type, tm t
...
+	if err := seed20.LoadAssertions(db, commitTo); err != nil {
+		return nil, nil, time.Time{}, err
+	}
+
+	// load and verify info about essential snaps
+	if err := seed20.LoadEssentialMeta(essentialTypes, tm); err != nil {
+		return nil, nil, time.Time{}, err
+	}
+
+	// consider model own timestamp as well
```
Comment (anonymouse64) [inline seed/seed.go:208]:
```suggestion
	// consider the model's timestamp as well - it must be signed
	// by the brand so is safe from the attack detailed above
```

File: asserts/batch_test.go [hunk 2]
```diff
@@ -103,6 +103,48 @@ func (s *batchSuite) TestAddStream(c *C) {
...
+
+	devAcct, err := s.db.Find(asserts.AccountType, map[string]string{
+		"account-id": s.dev1Acct.AccountID(),
+	})
+	c.Assert(err, IsNil)
+	c.Check(devAcct.(*asserts.Account).Username(), Equals, "developer1")
+
+	// this is the order they needed to be added
+	c.Check(seen, DeepEquals, []*asserts.Ref{
+
```
Comment (anonymouse64) [inline asserts/batch_test.go:142]:
```suggestion
```

File: seed/seed20_test.go [hunk 3]
```diff
@@ -832,6 +832,131 @@ func (s *seed20Suite) TestLoadEssentialMetaCore20(c *C) {
...
+			mod, essSnaps, betterTime, err := seed.ReadSystemEssentialAndBetterEarliestTime(s.SeedDir, sysLabel, t.onlyTypes, earliestTime, s.perfTimings)
+			c.Assert(err, IsNil)
+			c.Check(mod.BrandID(), Equals, "my-brand")
+			c.Check(mod.Model(), Equals, "my-model")
+			c.Check(mod.Timestamp().Equal(modelTime), Equals, true)
+			c.Check(essSnaps, HasLen, len(t.expected))
+			c.Check(essSnaps, DeepEquals, t.expected)
+			c.Check(betterTime.Equal(improvedTime), Equals, true, Commentf("%v expected: %v", betterTime, improvedTime))
+		}
+
```
Comment (anonymouse64) [inline seed/seed20_test.go:939]:
```suggestion
```

**Review by anonymouse64 — APPROVED** [review]
I think we should also do this for snap-declarations too, but other than that looks great to me (minus some nitpicky wording things in the doc-comments)

File: seed/seed.go [hunk 4]
```diff
@@ -153,3 +154,62 @@ func ReadSystemEssential(seedDir, label string, essentialTypes []snap.Type, tm t
...
+	if err != nil {
+		return nil, nil, time.Time{}, err
+
+	}
+
+	improve := func(a asserts.Assertion) {
+		// we consider only snap-revisions as they are stored-signed
+		// and more recent than snap-declarations anyway,
+		// other assertions are ignored as they might be added
+		// with unreliable times
```
[the remark this answers is not in the document]
Comment (pedronis) [reply seed/seed.go:180]: I used a slightly different explanation in the end

File: seed/seed.go [hunk 5]
```diff
@@ -153,3 +154,75 @@ func ReadSystemEssential(seedDir, label string, essentialTypes []snap.Type, tm t
...
+		return nil, nil, time.Time{}, fmt.Errorf("system label cannot be empty")
+	}
+	seed20, err := Open(seedDir, label)
+	if err != nil {
+		return nil, nil, time.Time{}, err
+
+	}
+
+	improve := func(a asserts.Assertion) {
+		// we consider only snap-revision and snap-declaration
```
Comment (mvo5) [inline seed/seed.go:177]: Thanks for this comment

File: seed/seed.go [hunk 6]
```diff
@@ -153,3 +154,75 @@ func ReadSystemEssential(seedDir, label string, essentialTypes []snap.Type, tm t
...
+		// assertions here as they must be store-signed, see
+		// checkConsistency for each type
+		// other assertions might be signed not by the store
+		// nor the brand so they might be provided by an
+		// attacker, signed using a registered key but
+		// containing unreliable time
+		var tstamp time.Time
+		switch a.Type() {
+		default:
```
Comment (mvo5) [inline seed/seed.go:186]: (nitpick^99) most of the switch/case I have seen has the default as the last entry but that is really probably my most nitpicky comment yet.

**Review by mvo5 — APPROVED** [review]
Thanks for this

**Pull request merged**